File: python/model_selector.py
from __future__ import annotations

from copilot import CopilotClient


async def select_model(client: CopilotClient, preferred_model: str) -> str | None:
    models = await client.list_models()
    if not models:
        return None

    default_index = next(
        (index for index, model in enumerate(models) if model.id.lower() == preferred_model.lower()),
        0,
    )

    print("Choose a Copilot model:")
    for index, model in enumerate(models, start=1):
        print(f"  {index}. {model.id}")

    selected_index = read_index(f"Model [{default_index + 1}]: ", len(models), default_index)
    return models[selected_index].id


def read_index(prompt: str, option_count: int, default_index: int) -> int:
    while True:
        answer = input(prompt).strip()
        if not answer:
            return default_index
        try:
            selection = int(answer)
        except ValueError:
            selection = 0
        if 1 <= selection <= option_count:
            return selection - 1
        print(f"Enter a number from 1 to {option_count}.")

File: nodejs/src/model-selector.ts
import type { CopilotClient } from "@github/copilot-sdk";
import { createInterface } from "node:readline/promises";
import { stdin as input, stdout as output } from "node:process";

export async function selectModel(client: CopilotClient, preferredModel: string): Promise<string | undefined> {
  const models = await client.listModels();
  if (models.length === 0) {
    return undefined;
  }

  const preferredIndex = models.findIndex((model) => model.id.toLowerCase() === preferredModel.toLowerCase());
  const defaultIndex = preferredIndex >= 0 ? preferredIndex : 0;

  console.log("Choose a Copilot model:");
  models.forEach((model, index) => console.log(`  ${index + 1}. ${model.id}`));

  const selectedIndex = await readIndex(`Model [${defaultIndex + 1}]: `, models.length, defaultIndex);
  return models[selectedIndex]?.id;
}

export async function readIndex(prompt: string, optionCount: number, defaultIndex: number): Promise<number> {
  const rl = createInterface({ input, output });
  try {
    while (true) {
      const answer = (await rl.question(prompt)).trim();
      if (!answer) {
        return defaultIndex;
      }

      const selection = Number.parseInt(answer, 10);
      if (Number.isInteger(selection) && selection >= 1 && selection <= optionCount) {
        return selection - 1;
      }

      console.log(`Enter a number from 1 to ${optionCount}.`);
    }
  } finally {
    rl.close();
  }
}
